### PR TITLE
New: Collapsible sections

### LINKED ISSUE #1436: Section overview on large design systems.
**The problem**

When documenting design systems, we want to include general design documentation and guidelines. In a DS documentation, component directory is just one major section. Eg.
1. Introduction
2. Design Principles
3. Tone of Voice
4. Images and illustrations
5. UI components

**Proposed solution**

Let's make it easier to split large design systems to independent sections. Either by enabling section collapsing or some kind of main navigation.

For example see
 http://digitalstyle.teliacompany.com/ (Collapsible sections)
and
  https://pulse.heartbeat.ua/  (main navigation)

**Alternative solutions**

Alternative solution to new feature would be example scripts, eg. for building multiple styleguidist instances, one for each section.

### PR DESCRIPTION
**Pull request by elevatebart**
closes #1436

- [X] Make sections collapsible if you ask for it
  - [X] tests
  - [x] Document new option `collapseSections`
- [X] Adjust types in `schema/config.ts` (very minor) for sections
- [X] Add comments on unclear global types and simplify them

### CODE REVIEW

File: docs/Configuration.md [hunk 1]
```diff
@@ -58,6 +58,12 @@ Type: `String` or `Array`, optional
 
 Your application static assets folder will be accessible as `/` in the style guide dev server.
 
+#### `collapseSections`
```
Comment (sapegin) [inline docs/Configuration.md]: Should we an API similar to `exampleMode` option with `expand` and `collapse` values?

Comment (elevatebart) [reply docs/Configuration.md]: Hello @sapegin, 
I think that `tocmode:'expand'` and `tocmode:'collapse'` could work. 
I am afraid of this being confusing to first-time users.
It is easier to misspell `collapse` that `true` or `false`. VScode or sublime text will give you the squiglies if you mistype.

Comment (sapegin) [reply docs/Configuration.md]: Yeah, but having different styles in the API is more confusing ;-)

Comment (elevatebart) [reply docs/Configuration.md]: `tocMode: 'expand'` then ?

Comment (sapegin) [reply docs/Configuration.md]: Yeah, it might not the most perfect API but at least consistent with the rest ;-)

Comment (elevatebart) [reply docs/Configuration.md]: I updated the docs as well

File: src/client/rsg-components/ComponentsList/ComponentsList.tsx [hunk 2]
```diff
@@ -16,20 +18,40 @@ const ComponentsList: React.FunctionComponent<ComponentsListProps> = ({
...
+				: getUrl({
+						name: item.name,
+						slug: item.slug,
+						anchor: !useRouterLinks,
+						hashPath: useRouterLinks ? hashPath : false,
+						id: useRouterLinks ? useHashId : false,
+				  });
+
+			const [open, setOpen] =
+				tocMode !== 'collapse' ? [true, () => {}] : React.useState(!!item.open);
```
Comment (sapegin) [inline src/client/rsg-components/ComponentsList/ComponentsList.tsx]: This looks like a violation of hooks rule, will it even work?

https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

File: src/client/rsg-components/TableOfContents/TableOfContents.spec.tsx [hunk 3]
```diff
@@ -192,9 +194,61 @@ it('should render components of a single top section as root', () => {
...
+		    "open": true,
 		    "sections": Array [],
 		    "slug": "textarea",
 		  },
 		]
 	`);
 });
+
+it('should render components with useRouterLinks', () => {
+	const actual = shallow(
```
Comment (sapegin) [inline src/client/rsg-components/TableOfContents/TableOfContents.spec.tsx]: Let's use RTL for new tests is possible ;-)

Comment (elevatebart) [reply src/client/rsg-components/TableOfContents/TableOfContents.spec.tsx]: good point, will do

File: src/client/rsg-components/ComponentsList/ComponentsList.tsx (hunk 2 above)
Comment (elevatebart) [reply src/client/rsg-components/ComponentsList/ComponentsList.tsx]: It does work I did not want to create sub components in there. I will.

File: src/client/rsg-components/TableOfContents/TableOfContents.spec.tsx (hunk 3 above)
Comment (elevatebart) [reply src/client/rsg-components/TableOfContents/TableOfContents.spec.tsx]: I tried my best to create this test and the only way I found using `data-test-id` in production.
Should I do that?

Comment (elevatebart) [reply src/client/rsg-components/TableOfContents/TableOfContents.spec.tsx]: scratch that I figured it out

File: src/client/rsg-components/ComponentsList/ComponentsList.tsx (hunk 2 above)
Comment (elevatebart) [reply src/client/rsg-components/ComponentsList/ComponentsList.tsx]: done

**Comment by elevatebart** [comment]
Hello @sapegin, do you think this Pull Request could be merged in?

File: src/client/rsg-components/ComponentsList/ComponentsListRenderer.tsx [hunk 4]
```diff
@@ -39,50 +38,55 @@ const styles = ({ color, fontFamily, fontSize, space, mq }: Rsg.Theme) => ({
...
+
+const ComponentsListSectionRenderer: React.FunctionComponent<Rsg.TOCItem & JssInjectedProps> = ({
+	classes,
+	heading,
+	visibleName,
+	href,
+	content,
+	shouldOpenInNewTab,
+	selected,
+	open: _open,
```
Comment (sapegin) [inline src/client/rsg-components/ComponentsList/ComponentsListRenderer.tsx]:
```suggestion
	open: initialOpen,
```

Let's use meaningful names.

File: src/client/rsg-components/TableOfContents/TableOfContents.spec.tsx [hunk 5]
```diff
@@ -176,25 +182,99 @@ it('should render components of a single top section as root', () => {
...
+
+	expect((getAllByRole('link')[0] as any).href).toMatch(/\/#\/Components$/);
+});
+
+/**
+ * testing this layer with no mocking makes no sense...
+ * This test should not exist but for good coverage policy this is necessary
+ */
+it('should detect sections containing current selection when tocMode', () => {
+	window.history.pushState({}, 'Collapse', 'http://localhost/#Button');
```
Comment (sapegin) [inline src/client/rsg-components/TableOfContents/TableOfContents.spec.tsx]: Dependency injection might be a better approach here.

File: src/client/rsg-components/ComponentsList/ComponentsListRenderer.tsx (hunk 4 above)
Comment (elevatebart) [reply src/client/rsg-components/ComponentsList/ComponentsListRenderer.tsx]: indeed !!

File: src/client/rsg-components/TableOfContents/TableOfContents.spec.tsx (hunk 5 above)
Comment (elevatebart) [reply src/client/rsg-components/TableOfContents/TableOfContents.spec.tsx]: This was a good idea. I used the same approach as `renderStyleguide.ts`

**Comment by elevatebart** [comment]
I took into account all of your comments.
@sapegin you have my green light to merge

**Pull request merged**